File: src/dpf.rs
// SPDX-License-Identifier: Apache-2.0
// Copyright (C) 2023 Yulong Ming (myl7)

//! See [`Dpf`].

use bitvec::prelude::*;
#[cfg(feature = "multi-thread")]
use rayon::prelude::*;

use crate::group::Group;
use crate::utils::{xor, xor_inplace};
pub use crate::PointFn;
use crate::{Cw, Prg, Share};

/// API of distributed point functions (DPFs).
///
/// `PointFn` used here means `$f(x) = \beta$` iff. `$x = \alpha$`, otherwise `$f(x) = 0$`.
///
/// - See [`PointFn`] for `IN_BLEN` and `OUT_BLEN`.
/// - See [`DpfImpl`] for the implementation.
pub trait Dpf<const IN_BLEN: usize, const OUT_BLEN: usize, G>
where
    G: Group<OUT_BLEN>,
{
    /// `s0s` is `$s^{(0)}_0$` and `$s^{(0)}_1$` which should be randomly sampled.
    fn gen(
        &self,
        f: &PointFn<IN_BLEN, OUT_BLEN, G>,
        s0s: [&[u8; OUT_BLEN]; 2],
    ) -> Share<OUT_BLEN, G>;

    /// `b` is the party. `false` is 0 and `true` is 1.
    fn eval(&self, b: bool, k: &Share<OUT_BLEN, G>, xs: &[&[u8; IN_BLEN]], ys: &mut [&mut G]);

    /// Full domain eval.
    /// See [`Dpf::eval`] for `b`.
    /// The corresponding `xs` to `ys` is the big endian representation of `0..=u*::MAX`.
    fn full_eval(&self, b: bool, k: &Share<OUT_BLEN, G>, ys: &mut [&mut G]);
}

/// Implementation of [`Dpf`].
pub struct DpfImpl<const IN_BLEN: usize, const OUT_BLEN: usize, P>
where
    P: Prg<OUT_BLEN, 1>,
{
    prg: P,
    filter_bitn: usize,
}

impl<const IN_BLEN: usize, const OUT_BLEN: usize, P> DpfImpl<IN_BLEN, OUT_BLEN, P>
where
    P: Prg<OUT_BLEN, 1>,
{
    pub fn new(prg: P) -> Self {
        Self {
            prg,
            filter_bitn: 8 * IN_BLEN,
        }
    }

    pub fn new_with_filter(prg: P, filter_bitn: usize) -> Self {
        assert!(filter_bitn <= 8 * IN_BLEN && filter_bitn > 1);
        Self { prg, filter_bitn }
    }
}

const IDX_L: usize = 0;
const IDX_R: usize = 1;

impl<const IN_BLEN: usize, const OUT_BLEN: usize, P, G> Dpf<IN_BLEN, OUT_BLEN, G>
    for DpfImpl<IN_BLEN, OUT_BLEN, P>
where
    P: Prg<OUT_BLEN, 1>,
    G: Group<OUT_BLEN>,
{
    fn gen(
        &self,
        f: &PointFn<IN_BLEN, OUT_BLEN, G>,
        s0s: [&[u8; OUT_BLEN]; 2],
    ) -> Share<OUT_BLEN, G> {
        // The bit size of `$\alpha$`.
        let n = self.filter_bitn;
        // Set `$s^{(1)}_0$` and `$s^{(1)}_1$`.
        let mut ss_prev = [s0s[0].to_owned(), s0s[1].to_owned()];
        // Set `$t^{(0)}_0$` and `$t^{(0)}_1$`.
        let mut ts_prev = [false, true];
        let mut cws = Vec::<Cw<OUT_BLEN, G>>::with_capacity(n);
        for i in 0..n {
            // MSB is required since we index from high to low in arrays.
            let alpha_i = f.alpha.view_bits::<Msb0>()[i];
            let [([s0l], t0l), ([s0r], t0r)] = self.prg.gen(&ss_prev[0]);
            let [([s1l], t1l), ([s1r], t1r)] = self.prg.gen(&ss_prev[1]);
            let (keep, lose) = if alpha_i {
                (IDX_R, IDX_L)
            } else {
                (IDX_L, IDX_R)
            };
            let s_cw = xor(&[[&s0l, &s0r][lose], [&s1l, &s1r][lose]]);
            let tl_cw = t0l ^ t1l ^ alpha_i ^ true;
            let tr_cw = t0r ^ t1r ^ alpha_i;
            let cw = Cw {
                s: s_cw,
                v: G::zero(),
                tl: tl_cw,
                tr: tr_cw,
            };
            cws.push(cw);
            ss_prev = [
                xor(&[
                    [&s0l, &s0r][keep],
                    if ts_prev[0] { &s_cw } else { &[0; OUT_BLEN] },
                ]),
                xor(&[
                    [&s1l, &s1r][keep],
                    if ts_prev[1] { &s_cw } else { &[0; OUT_BLEN] },
                ]),
            ];
            ts_prev = [
                [t0l, t0r][keep] ^ (ts_prev[0] & [tl_cw, tr_cw][keep]),
                [t1l, t1r][keep] ^ (ts_prev[1] & [tl_cw, tr_cw][keep]),
            ];
        }
        let cw_np1 =
            (f.beta.clone() + -Into::<G>::into(ss_prev[0]) + ss_prev[1].into()).neg_if(ts_prev[1]);
        Share {
            s0s: vec![s0s[0].to_owned(), s0s[1].to_owned()],
            cws,
            cw_np1,
        }
    }

    fn eval(&self, b: bool, k: &Share<OUT_BLEN, G>, xs: &[&[u8; IN_BLEN]], ys: &mut [&mut G]) {
        #[cfg(feature = "multi-thread")]
        self.eval_mt(b, k, xs, ys);
        #[cfg(not(feature = "multi-thread"))]
        self.eval_st(b, k, xs, ys);
    }

    fn full_eval(&self, b: bool, k: &Share<OUT_BLEN, G>, ys: &mut [&mut G]) {
        let n = k.cws.len();
        assert_eq!(n, self.filter_bitn);

        let s = k.s0s[0];
        let t = b;
        self.full_eval_layer(b, k, ys, 0, (s, t));
    }
}

impl<const IN_BLEN: usize, const OUT_BLEN: usize, P> DpfImpl<IN_BLEN, OUT_BLEN, P>
where
    P: Prg<OUT_BLEN, 1>,
{
    /// Eval with single-threading.
    /// See [`Dpf::eval`].
    pub fn eval_st<G>(
        &self,
        b: bool,
        k: &Share<OUT_BLEN, G>,
        xs: &[&[u8; IN_BLEN]],
        ys: &mut [&mut G],
    ) where
        G: Group<OUT_BLEN>,
    {
        xs.iter()
            .zip(ys.iter_mut())
            .for_each(|(x, y)| self.eval_point(b, k, x, y));
    }

    #[cfg(feature = "multi-thread")]
    /// Eval with multi-threading.
    /// See [`Dpf::eval`].
    pub fn eval_mt<G>(
        &self,
        b: bool,
        k: &Share<OUT_BLEN, G>,
        xs: &[&[u8; IN_BLEN]],
        ys: &mut [&mut G],
    ) where
        G: Group<OUT_BLEN>,
    {
        xs.par_iter()
            .zip(ys.par_iter_mut())
            .for_each(|(x, y)| self.eval_point(b, k, x, y));
    }

    fn full_eval_layer<G>(
        &self,
        b: bool,
        k: &Share<OUT_BLEN, G>,
        ys: &mut [&mut G],
        layer_i: usize,
        (s, t): ([u8; OUT_BLEN], bool),
    ) where
        G: Group<OUT_BLEN>,
    {
        assert_eq!(ys.len(), 1 << (self.filter_bitn - layer_i));
        if ys.len() == 1 {
            *ys[0] = (Into::<G>::into(s) + if t { k.cw_np1.clone() } else { G::zero() }).neg_if(b);
            return;
        }

        let cw = &k.cws[layer_i];
        let [([mut sl], mut tl), ([mut sr], mut tr)] = self.prg.gen(&s);
        xor_inplace(&mut sl, &[if t { &cw.s } else { &[0; OUT_BLEN] }]);
        xor_inplace(&mut sr, &[if t { &cw.s } else { &[0; OUT_BLEN] }]);
        tl ^= t & cw.tl;
        tr ^= t & cw.tr;

        let (ys_l, ys_r) = ys.split_at_mut(ys.len() / 2);
        #[cfg(feature = "multi-thread")]
        rayon::join(
            || self.full_eval_layer(b, k, ys_l, layer_i + 1, (sl, tl)),
            || self.full_eval_layer(b, k, ys_r, layer_i + 1, (sr, tr)),
        );
        #[cfg(not(feature = "multi-thread"))]
        {
            self.full_eval_layer(b, k, ys_l, layer_i + 1, (sl, tl));
            self.full_eval_layer(b, k, ys_r, layer_i + 1, (sr, tr));
        }
    }

    pub fn eval_point<G>(&self, b: bool, k: &Share<OUT_BLEN, G>, x: &[u8; IN_BLEN], y: &mut G)
    where
        G: Group<OUT_BLEN>,
    {
        let n = k.cws.len();
        assert_eq!(n, self.filter_bitn);
        let v = y;

        let mut s_prev = k.s0s[0];
        let mut t_prev = b;
        for i in 0..n {
            let cw = &k.cws[i];
            let [([mut sl], mut tl), ([mut sr], mut tr)] = self.prg.gen(&s_prev);
            xor_inplace(&mut sl, &[if t_prev { &cw.s } else { &[0; OUT_BLEN] }]);
            xor_inplace(&mut sr, &[if t_prev { &cw.s } else { &[0; OUT_BLEN] }]);
            tl ^= t_prev & cw.tl;
            tr ^= t_prev & cw.tr;
            if x.view_bits::<Msb0>()[i] {
                s_prev = sr;
                t_prev = tr;
            } else {
                s_prev = sl;
                t_prev = tl;
            }
        }
        *v =
            (Into::<G>::into(s_prev) + if t_prev { k.cw_np1.clone() } else { G::zero() }).neg_if(b);
    }
}

#[cfg(all(test, feature = "prg"))]
mod tests {
    use arbtest::arbtest;
    use std::iter;

    use super::*;
    use crate::group::byte::ByteGroup;
    use crate::prg::Aes128MatyasMeyerOseasPrg;

    type GroupImpl = ByteGroup<16>;
    type PrgImpl = Aes128MatyasMeyerOseasPrg<16, 1, 2>;
    type DpfImplImpl = DpfImpl<2, 16, PrgImpl>;

    #[test]
    fn test_correctness() {
        arbtest(|u| {
            let keys: [[u8; 16]; 2] = u.arbitrary()?;
            let prg = PrgImpl::new(&std::array::from_fn(|i| &keys[i]));
            let filter_bitn = u.arbitrary::<usize>()? % 15 + 2; // 2..=16
            let dpf = DpfImplImpl::new_with_filter(prg, filter_bitn);
            let s0s: [[u8; 16]; 2] = u.arbitrary()?;
            let alpha_i: u16 = u.arbitrary::<u16>()? >> (16 - filter_bitn);
            let alpha: [u8; 2] = (alpha_i << (16 - filter_bitn)).to_be_bytes();
            let beta: [u8; 16] = u.arbitrary()?;
            let f = PointFn {
                alpha,
                beta: beta.into(),
            };
            let k = dpf.gen(&f, [&s0s[0], &s0s[1]]);
            let mut k0 = k.clone();
            k0.s0s = vec![k0.s0s[0]];
            let mut k1 = k;
            k1.s0s = vec![k1.s0s[1]];

            let xs: Vec<_> = (0u16..=u16::MAX >> (16 - filter_bitn))
                .map(|i| (i << (16 - filter_bitn)).to_be_bytes())
                .collect();
            assert_eq!(xs.len(), 1 << filter_bitn);
            let xs_lt_num = alpha_i;
            let xs_gt_num = (u16::MAX >> (16 - filter_bitn)) - alpha_i;
            let ys_expected: Vec<_> = iter::repeat(beta.into())
                .take(xs_lt_num as usize)
                .chain([beta.into()])
                .chain(iter::repeat(GroupImpl::zero()).take(xs_gt_num as usize))
                .collect();

            let mut ys0 = vec![GroupImpl::zero(); xs.len()];
            let mut ys1 = vec![GroupImpl::zero(); xs.len()];
            dpf.eval(
                false,
                &k0,
                &xs.iter().collect::<Vec<_>>(),
                &mut ys0.iter_mut().collect::<Vec<_>>(),
            );
            ys0.iter().for_each(|y0| {
                assert_ne!(*y0, GroupImpl::zero());
                assert_ne!(*y0, [0xff; 16].into());
            });
            dpf.eval(
                true,
                &k1,
                &xs.iter().collect::<Vec<_>>(),
                &mut ys1.iter_mut().collect::<Vec<_>>(),
            );
            ys1.iter().for_each(|y1| {
                assert_ne!(*y1, GroupImpl::zero());
                assert_ne!(*y1, [0xff; 16].into());
            });
            let ys: Vec<_> = ys0
                .iter()
                .zip(ys1.iter())
                .map(|(y0, y1)| y0.clone() + y1.clone())
                .collect();
            assert_ys_eq(&ys, &ys_expected, &xs, &alpha);

            let mut ys0_full_eval = vec![ByteGroup::zero(); 1 << filter_bitn];
            dpf.full_eval(
                false,
                &k0,
                &mut ys0_full_eval.iter_mut().collect::<Vec<_>>(),
            );
            assert_ys_eq(&ys0_full_eval, &ys0, &xs, &alpha);
            let mut ys1_full_eval = vec![ByteGroup::zero(); 1 << filter_bitn];
            dpf.full_eval(true, &k1, &mut ys1_full_eval.iter_mut().collect::<Vec<_>>());
            assert_ys_eq(&ys1_full_eval, &ys1, &xs, &alpha);

            Ok(())
        });
    }

    fn assert_ys_eq(ys: &[GroupImpl], ys_expected: &[GroupImpl], xs: &[[u8; 2]], alpha: &[u8; 2]) {
        let alpha_int = u16::from_be_bytes(*alpha);
        for (i, (x, (y, y_expected))) in
            xs.iter().zip(ys.iter().zip(ys_expected.iter())).enumerate()
        {
            let x_int = u16::from_be_bytes(*x);
            let cmp = if x_int < alpha_int {
                "<"
            } else if x_int > alpha_int {
                ">"
            } else {
                "="
            };
            assert_eq!(y, y_expected, "where i={}, x={:?}, x{}alpha", i, *x, cmp);
        }
    }
}

File: src/dpf.c
// SPDX-License-Identifier: Apache-2.0
// Copyright (C) 2025 Yulong Ming (myl7)

#include <fss/dpf.h>
#include <string.h>
#include "utils.h"

// Load the 1bit t from MSB, so we can truncate during adding
FSS_CUDA_HOST_DEVICE static inline void load_st(uint8_t *s, uint8_t *t) {
  *t = get_bit_lsb(s, kLambda * 8 - 1);
  set_bit_lsb(s, kLambda * 8 - 1, 0);
}

static inline void set_st(uint8_t *s, uint8_t t) {
  set_bit_lsb(s, kLambda * 8 - 1, t);
}

FSS_CUDA_HOST_DEVICE static inline void load_sst(uint8_t *ss, uint8_t *t0, uint8_t *t1) {
  load_st(ss, t0);
  load_st(ss + kLambda, t1);
}

// Save the 2bit tl tr in an extra byte
FSS_CUDA_HOST_DEVICE static inline void set_cwt(uint8_t *cw, uint8_t tl, uint8_t tr) {
  cw[kLambda] = tl << 1 | tr;
}

FSS_CUDA_HOST_DEVICE static inline void get_cwt(const uint8_t *cw, uint8_t *tl, uint8_t *tr) {
  *tl = cw[kLambda] >> 1;
  *tr = cw[kLambda] & 1;
}

// | s0 | s1 | s0l | s0r | s1l | s1r |
// | ss      | s0s      | s1s        |
FSS_CUDA_HOST_DEVICE void dpf_gen(Key k, PointFunc pf, uint8_t *sbuf) {
  uint8_t *ss = sbuf;
  uint8_t *s0 = ss;
  uint8_t *s1 = ss + kLambda;
  uint8_t t0, t1;
  load_sst(ss, &t0, &t1);
  t0 = 0;
  t1 = 1;
  Point p = pf.point;

  uint8_t *s0s = sbuf + kLambda * 2;
  uint8_t *s0l = s0s;
  uint8_t *s0r = s0s + kLambda;
  uint8_t *s1s = sbuf + kLambda * 4;
  uint8_t *s1l = s1s;
  uint8_t *s1r = s1s + kLambda;
  uint8_t t0l, t0r, t1l, t1r;

  for (int i = 0; i < p.alpha.bitlen; i++) {
    prg(s0s, 2 * kLambda, s0);
    prg(s1s, 2 * kLambda, s1);
    load_sst(s0s, &t0l, &t0r);
    load_sst(s1s, &t1l, &t1r);

    // Actually get MSB first
    uint8_t alpha_i = get_bit_lsb(p.alpha.bytes, p.alpha.bitlen - i - 1);
    uint8_t *cw = k.cws + i * kDpfCwLen;

    uint8_t *s0_lose = alpha_i ? s0l : s0r;
    uint8_t *s1_lose = alpha_i ? s1l : s1r;

    uint8_t *s_cw = cw;
    memcpy(s_cw, s0_lose, kLambda);
    xor_bytes(s_cw, s1_lose, kLambda);

    uint8_t tl_cw, tr_cw;
    tl_cw = t0l ^ t1l ^ alpha_i ^ 1;
    tr_cw = t0r ^ t1r ^ alpha_i;
    set_cwt(cw, tl_cw, tr_cw);

    uint8_t *s0_keep = alpha_i ? s0r : s0l;
    uint8_t *s1_keep = alpha_i ? s1r : s1l;
    uint8_t t0_keep = alpha_i ? t0r : t0l;
    uint8_t t1_keep = alpha_i ? t1r : t1l;
    uint8_t t_cw_keep = alpha_i ? tr_cw : tl_cw;

    memcpy(s0, s0_keep, kLambda);
    if (t0) xor_bytes(s0, s_cw, kLambda);
    memcpy(s1, s1_keep, kLambda);
    if (t1) xor_bytes(s1, s_cw, kLambda);

    if (t0) t0 = t0_keep ^ t_cw_keep;
    else t0 = t0_keep;
    if (t1) t1 = t1_keep ^ t_cw_keep;
    else t1 = t1_keep;
  }

  memcpy(k.cw_np1, p.beta, kLambda);
  set_bit_lsb(k.cw_np1, kLambda * 8 - 1, 0);
  group_neg(s0);
  group_add(k.cw_np1, s0);
  group_add(k.cw_np1, s1);
  if (t1) group_neg(k.cw_np1);
}

// | s | sl | sr |
// |   | ss      |
FSS_CUDA_HOST_DEVICE void dpf_eval(uint8_t *sbuf, uint8_t b, Key k, Bits x) {
  uint8_t *s = sbuf;
  uint8_t t;
  load_st(s, &t);
  t = b;

  uint8_t *ss = sbuf + kLambda;
  uint8_t *sl = ss;
  uint8_t *sr = ss + kLambda;
  uint8_t tl, tr;

  for (int i = 0; i < x.bitlen; i++) {
    const uint8_t *cw = k.cws + i * kDpfCwLen;
    const uint8_t *s_cw = cw;
    uint8_t tl_cw, tr_cw;
    get_cwt(cw, &tl_cw, &tr_cw);

    prg(ss, 2 * kLambda, s);
    load_sst(ss, &tl, &tr);
    if (t) {
      xor_bytes(sl, s_cw, kLambda);
      xor_bytes(sr, s_cw, kLambda);
      tl ^= tl_cw;
      tr ^= tr_cw;
    }

    // Actually get MSB first
    uint8_t x_i = get_bit_lsb(x.bytes, x.bitlen - i - 1);

    memcpy(s, x_i ? sr : sl, kLambda);
    t = x_i ? tr : tl;
  }

  if (t) group_add(s, k.cw_np1);
  if (b) group_neg(s);
}

#include <assert.h>
#include <stdlib.h>
#include <omp.h>

void dpf_eval_full_domain_node(int depth, uint8_t *sbufl, uint8_t *sbufr, uint8_t b, Key k) {
  uint8_t *s = sbufl;
  uint8_t t;
  load_st(s, &t);

  uint8_t *ss = (uint8_t *)malloc(kLambda * 2);
  assert(ss != NULL);
  uint8_t *sl = ss;
  uint8_t *sr = ss + kLambda;
  uint8_t tl, tr;

  const uint8_t *cw = k.cws + depth * kDpfCwLen;
  const uint8_t *s_cw = cw;
  uint8_t tl_cw, tr_cw;
  get_cwt(cw, &tl_cw, &tr_cw);

  prg(ss, 2 * kLambda, s);
  load_sst(ss, &tl, &tr);
  if (t) {
    xor_bytes(sl, s_cw, kLambda);
    xor_bytes(sr, s_cw, kLambda);
    tl ^= tl_cw;
    tr ^= tr_cw;
  }

  memcpy(sbufl, sl, kLambda);
  set_st(sbufl, tl);
  memcpy(sbufr, sr, kLambda);
  set_st(sbufr, tr);
  free(ss);
}

void dpf_eval_full_domain_leaf(uint8_t *sbuf, uint8_t b, Key k) {
  uint8_t *s = sbuf;
  uint8_t t;
  load_st(s, &t);

  if (t) group_add(s, k.cw_np1);
  if (b) group_neg(s);
}

void dpf_eval_full_domain_subtree(
  int depth, uint8_t *sbuf, size_t l, size_t r, uint8_t b, Key k, int x_bitlen, int par_depth) {
  assert(kLambda * (1ULL << (x_bitlen - depth)) == r - l);

  if (depth == x_bitlen) {
    dpf_eval_full_domain_leaf(sbuf + l, b, k);
    return;
  }

  size_t mid = (l + r) / 2;
  dpf_eval_full_domain_node(depth, sbuf + l, sbuf + mid, b, k);

  if (depth < par_depth) {
#pragma omp parallel
#pragma omp single
    {
#pragma omp task
      { dpf_eval_full_domain_subtree(depth + 1, sbuf, l, mid, b, k, x_bitlen, par_depth); }
#pragma omp task
      { dpf_eval_full_domain_subtree(depth + 1, sbuf, mid, r, b, k, x_bitlen, par_depth); }
#pragma omp taskwait
    }
  } else {
    dpf_eval_full_domain_subtree(depth + 1, sbuf, l, mid, b, k, x_bitlen, par_depth);
    dpf_eval_full_domain_subtree(depth + 1, sbuf, mid, r, b, k, x_bitlen, par_depth);
  }
}

void dpf_eval_full_domain(uint8_t *sbuf, uint8_t b, Key k, int x_bitlen) {
  uint8_t *s = sbuf;
  uint8_t t = b;
  set_st(s, t);

  int threads = omp_get_max_threads();
  int par_depth = 0;
  while ((1 << par_depth) <= threads) {
    par_depth++;
  }
  par_depth--;

  size_t sbuf_len = kLambda * (1ULL << x_bitlen);
  dpf_eval_full_domain_subtree(0, sbuf, 0, sbuf_len, b, k, x_bitlen, par_depth);
}
